pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1051 @@
+			InvestmentState::<T>::mutate(who, investment_id, |current_invest_state| {
+				// Should never occur but let's be safe
+				if let Some(InvestState::NoState) = maybe_invest_state {
+					*current_invest_state = None;
+				} else if maybe_invest_state != *current_invest_state {
+					*current_invest_state = maybe_invest_state;
+				}
+			});
+
+			// Need to check if `SwapReturnDone` is part of inner state without
+			// `ActiveSwapIntoReturnCurrency` as this implies the successful termination of
+			// a collect (with swap into return currency). If this is the case, the returned
+			// redeem state needs to be updated as well.
+			match maybe_redeem_state {
+				Some(RedeemState::InvestedAnd { inner, .. })
+				| Some(RedeemState::NotInvestedAnd { inner }) => {
+					let redeem_state = maybe_redeem_state.unwrap_or_default();
+					let maybe_collected_redeem_state = Self::apply_collect_redeem_transition(
+						who,
+						investment_id,
+						redeem_state,
+						inner,
+					)?;
+					Ok((maybe_invest_state, maybe_collected_redeem_state))
+				}
+				Some(_) => {
+					RedemptionState::<T>::mutate(who, investment_id, |current_redeem_state| {
+						// Update and emit event on mismatch
+						if maybe_redeem_state != *current_redeem_state {
+							*current_redeem_state = maybe_redeem_state;
+						}
+					});
+					Ok((maybe_invest_state, maybe_redeem_state))
+				}
+				None => Ok((maybe_invest_state, maybe_redeem_state)),
+			}
+			.map_err(|e: DispatchError| e)
+		}
+		// Update to provided value, if not none
+		else if let Some(swap_order) = maybe_swap {
+			Self::place_swap_order(who, investment_id, swap_order, reason)?;
+			Ok((None, None))
+		} else {
+			Ok((None, None))
+		}
+	}
+
+	/// Kills all storage associated with token swaps and cancels the
+	/// potentially active swap order.
+	///
+	/// NOTE: Must only be called in `handle_swap_order`.
+	fn kill_swap_order(who: &T::AccountId, investment_id: T::InvestmentId) -> DispatchResult {
+		if let Some(swap_order_id) = TokenSwapOrderIds::<T>::take(who, investment_id) {
+			T::TokenSwaps::cancel_order(swap_order_id)?;
+			ForeignInvestmentInfo::<T>::remove(swap_order_id);
+		}
+		Ok(())
+	}
+
+	/// Sets up `TokenSwapOrderIds` and `ForeignInvestmentInfo` storages, if the
+	/// order does not exist yet.
+	///
+	/// NOTE: Must only be called in `handle_swap_order`.
+	#[transactional]
+	fn place_swap_order(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		swap: SwapOf<T>,
+		reason: TokenSwapReason,
+	) -> DispatchResult {
+		// exit early
+		if swap.amount.is_zero() {
+			return Self::kill_swap_order(who, investment_id);
+		}
+		match TokenSwapOrderIds::<T>::get(who, investment_id) {
+			Some(swap_order_id) if T::TokenSwaps::is_active(swap_order_id) => {
+				T::TokenSwaps::update_order(
+					who.clone(),
+					swap_order_id,
+					swap.amount,
+					T::DefaultTokenSwapSellPriceLimit::get(),
+					T::DefaultTokenMinFulfillmentAmount::get(),
```

Comment:
I'm pretty sure we can not use these constants here in this way. At least "sell price limit" is related to the amount of the order. It's a protection againts a big fluctuation of that amount. i.e. I do not want to sell if for buying X I'm selling and amount under an Y.

Maybe @thea-leake can explain more here

Reply:
Thanks for pointing this out. I wanted to discuss this as part of the review as I realized quite early it would not work and then discovered these parameters to receive some changes in #1494

Reply:
The recent changes in #1494 made the default configuration for swaps through Foreign Investments a bit simpler. I have set the `MinFulfillmentAmount` to the exact swapping amount of the outgoing currency as we want everything to be swapped. Then, there is also the `SellRatio` which reflects the relative cost of the incoming currency to the outgoing one. So for stables, this should be exactly one. 

This one should always match the conversion rate introduced in this pallet which is set to `Rate::one()` as well for now. Improving it further, should be handled in another PR and is not relevant for the MVP.